### PR TITLE
Voice cross stave support in System

### LINKED ISSUE #1373: Design issue: Notes in a voice cannot be placed on different staves (using System)
This is not allowed by VexFlow and it seems that the solution is to make two voices out of one adding GhostNotes. However this is an impediment to have slurs or beams between staves (see #434)

I am collaborating with @jaredjj3 on https://github.com/stringsync/vexml. This is the application that I am using to produce the jsfiddle examples in the OSMD issues from musicxml files.

There I am facing currently this design issue with **test_clef_end_measure.xml** from OSMD

@AaronDavidNewman once vexml is more mature we will be able to generate real scores for testing. As you commented once it would be good to have more real score testing. Or as @mscuthbert did use http://lilypond.org/doc/v2.20/input/regression/musicxml/collated-files.html

@0xfe Do we want to make a design change to allow notes from one voice on different staves?

### PR DESCRIPTION
**Pull request by rvilarl**
fixes #1373
ready for review now with more focus. It includes the tests from @AaronDavidNewman

### CODE REVIEW

**Comment by rvilarl** [comment]
CrossBeam Single_clef_mixed_1 Bravura
CrossBeam Single_clef_mixed_2 Bravura
CrossBeam Mixed_clef_voice_middle Bravura
CrossBeam Vertical_alignment___cross_stave__beam_down1_ Bravura
CrossBeam Vertical_alignment___cross_stave__beam_down2_ Bravura
CrossBeam Vertical_alignment___cross_stave__beam_down3_ Bravura
CrossBeam Vertical_alignment___cross_stave__beam_down4_ Bravura
CrossBeam Vertical_alignment___cross_stave__beam_middle1_ Bravura
CrossBeam Vertical_alignment___cross_stave__beam_middle2_ Bravura
CrossBeam Vertical_alignment___cross_stave__beam_middle3_ Bravura
CrossBeam Vertical_alignment___cross_stave__beam_middle4_ Bravura
CrossBeam Vertical_alignment___cross_stave__beam_up1_ Bravura
CrossBeam Vertical_alignment___cross_stave__beam_up2_ Bravura
CrossBeam Vertical_alignment___cross_stave__beam_up3_ Bravura
CrossBeam Vertical_alignment___cross_stave__beam_up4_ Bravura

**Comment by rvilarl** [comment]
Ready for review

**Comment by 0xfe** [comment]
Thanks Rodrigo. This is good start. Re: the first two examples, are they intentionally off?

**Comment by AaronDavidNewman** [comment]
I wouldn't say 'intentionally' :)  But this is the state of mixed-direction beam currently.  Vex flow has always allowed mixed direction on a single staff, but there are no test cases for it.  It seems logical to include them with the cross-beam stuff.

Additional bugs/PRs can be opened against the those issues.  I would expect more than one change will be required.

**Comment by sschmidTU** [comment]
@0xfe Rodrigo mentioned in an earlier PR that these stem mismatch issues are outside the scope of the PR, will be handled separately.

**Comment by rvilarl** [comment]
This is right. The aim of this PR is simply to allow System to work with a Voice whose Notes are located on different Staves. The tests are there to show the issues related with the representation of these Voices. These issues will have to be solved in future PRs.

**Comment by rvilarl** [comment]
> Thanks Rodrigo. This is good start.

@0xfe Well this is not really a start, this started 5mo ago with #1373 there was also anther previous PR. It would be nice to get it through. I am happy to make changes if required.

**Comment by rvilarl** [comment]
Let us progress with this one this weekend :muscle: and make a new release :partying_face: !

File: src/voice.ts [hunk 1]
```diff
@@ -159,7 +159,7 @@ export class Voice extends Element {
       let boundingBox = undefined;
       for (let i = 0; i < this.tickables.length; ++i) {
         const tickable = this.tickables[i];
-        tickable.setStave(stave);
+        if (!tickable.getStave()) tickable.setStave(stave);
```
Comment (ronyeh) [inline src/voice.ts:162]: Just curious about this line. Did it fix a bug? Is there something wrong with blindly calling .setStave() every time? :-)

Comment (rvilarl) [reply src/voice.ts:162]: Ron thanks for the review! The tickabless of a voice can be located on different staves. The voice has a defect stave, The tickables should only be set to the default if they have no previously associated stave.

**Comment by rvilarl** [comment]
I decided to make a minimal change to Beam based on the work done by @Silverwolf90 in #97 to get a better representation of the cross beams. I hope that this improves the acceptance of this PR :)

**Comment by rvilarl** [comment]
CrossBeam Single_clef_mixed_1 Bravura
CrossBeam Single_clef_mixed_2 Bravura
CrossBeam Mixed_clef_voice_middle Bravura
CrossBeam Vertical_alignment___cross_stave__beam_up1_ Bravura
CrossBeam Vertical_alignment___cross_stave__beam_up2_ Bravura
CrossBeam Vertical_alignment___cross_stave__beam_up3_ Bravura
CrossBeam Vertical_alignment___cross_stave__beam_up4_ Bravura
CrossBeam Vertical_alignment___cross_stave__beam_down1_ Bravura
CrossBeam Vertical_alignment___cross_stave__beam_down2_ Bravura
CrossBeam Vertical_alignment___cross_stave__beam_down3_ Bravura
CrossBeam Vertical_alignment___cross_stave__beam_down4_ Bravura
CrossBeam Vertical_alignment___cross_stave__beam_middle1_ Bravura
CrossBeam Vertical_alignment___cross_stave__beam_middle2_ Bravura
CrossBeam Vertical_alignment___cross_stave__beam_middle3_ Bravura
CrossBeam Vertical_alignment___cross_stave__beam_middle4_ Bravura

**Comment by rvilarl** [comment]
Ready for review :)

File: src/voice.ts (hunk 1 above)
Comment (rvilarl) [reply src/voice.ts:162]: The aim of this PR is to allow Voices have Tickables in several Staves and to make sure that System also allows it.

**Comment by rvilarl** [comment]
@AaronDavidNewman I believe that you have tested already the system.ts related changes of this PR. Any issues? Did you realise that I also changed beam.ts fixing the stem related rendering problems? See outputs above.

**Comment by AaronDavidNewman** [comment]
no I haven't been paying attention.  I'll try to add some of the cases from the other issue, if that's what you need.  Def. looks better with the beaming change.

I haven't spent any additional time with system.ts but I haven't been able to find a better way to do it with a single structure.  We need to keep track of which notes are on which stave somehow.

**Comment by rvilarl** [comment]
> no I haven't been paying attention. I'll try to add some of the cases from the other issue, if that's what you need. Def. looks better with the beaming change.

Hi Aaron, just add additional tests, if you think that they would help to find issues.
 
> I haven't spent any additional time with system.ts but I haven't been able to find a better way to do it with a single structure. We need to keep track of which notes are on which stave somehow.

What else do we need? The notes already store the stave that they are connected to.

**Comment by rvilarl** [comment]
@0xfe is there a problem with this PR? I have the feeling that I have not been able to explain properly its aim.

**Comment by AaronDavidNewman** [comment]
@0xfe I think we should merge it.  It doesn't break anything, it gives us some additional capabilities.  I don't think it introduces any unnecessary complexity in the code or the interface.  So....why not?

I've already merged the change into Smoosic branch, but I haven't figured out how to work x-stave in the Smoosic UI yet.

**Comment by rvilarl** [comment]
@AaronDavidNewman I contacted Mohit by email and he confirmed that with your go I can merge, let us do it!

**Pull request merged**